File: driver/python/message.py
import json, socket

class MessageEncoder(json.JSONEncoder):
	def default(self, m):
		if isinstance(m, Message):
			return {'header': m.header, 'body': m.body}
		return json.JSONEncoder.default(self, obj)

class Message:
	ACKNOWLEDGE = 1
	CONNECTION_ESTABLISHED = 2
	FUNCTION_REGISTER = 3
	WORK_REQUEST = 4
	NO_WORK = 5
	WAKE = 6
	RUN_WORK = 7
	WORK_CREATED = 8
	WORK_RUNNING = 9
	WORK_STATUS = 10
	WORK_DONE = 11

	def __init__(self, code, body={}):
		self.header = {'code': code, 'origin': socket.gethostname()}
		self.body = body;
	
	def getCode(self):
		return self.header['code']
	
	def toJson(self):
		return json.dumps(self, cls=MessageEncoder);
	
	@staticmethod	
	def fromJson(data):
		dct = json.loads(data);
		if 'body' not in dct:
			dct['body'] = {}
		if 'header' not in dct:
			dct['header'] = {}
			if 'code' not in dct['header']:
				dct['header']['code'] = 0;
		return Message(dct['header']['code'], dct['body'])

File: driver/php/Message.php
<?php

namespace gtsd;

class Message {
    public $header;
    public $body;
    
    const ACKNOWLEDGE = 1;
    const CONNECTION_ESTABLISHED = 2;
    const FUNCTION_REGISTER = 3;
    const WORK_REQUEST = 4;
    const NO_WORK = 5;
    const WAKE = 6;
    const RUN_WORK = 7;
    const WORK_CREATED = 8;
    const WORK_RUNNING = 9;
    const WORK_STATUS = 10;
    const WORK_DONE = 11;
    
    public function __construct($code, $body = array(), $hostname = null)
    {
        if($hostname === null) $hostname = gethostname();
        
        $this->header = array('origin' => $hostname, 'code' => $code);
        $this->body = $body;
    }
    
    public function getCode()
    {
        return $this->header['code'];
    }
    
    public function getOrigin()
    {
        return $this->header['origin'];
    }
    
    public function getBody()
    {
        return $this->body;
    }
    
    public function getBodyData($data)
    {
        if(isset($this->body[$data]))
            return $this->body[$data];
        else
            return null;
    }
    
    public function __toString()
    {
        return json_encode(array('header' => $this->header, 'body' => $this->body));
    }
    
    public function dump()
    {
        echo "======================================================================\n";
        $content = "Code: ".$this->header['code']." | Origin: ".$this->header['origin'];
        echo "| ".$content;
        for($i=strlen($content)+3; $i<70; $i++) echo " ";
        echo "|\n";
        echo "======================================================================\n";
        foreach($this->body as $k=>$v)
        { 
            $cv = (is_array($v)) ? str_replace("\n", " ", var_export($v, 1)) : $v;
            $content = $k.": ".$cv;
            echo "| ".$content;
            for($i=strlen($content)+3; $i<70; $i++) echo " ";
            echo "|\n";
        }
        echo "======================================================================\n\n";
        
    }
    
    public static function fromJSON($data)
    {
        return new Message($data['header']['code'], $data['body'], $data['header']['origin']);
    }
}
